File: jAutoInvoice/src/java/net/sf/jautoinvoice/server/model/Modelo.java
/*
 * Modelo.java
 *
 * This file is part of jAutoInvoice, http://sourceforge.net/p/jautoinvoice
 *
 * Copyright (C) 2011 jAutoInvoice
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package net.sf.jautoinvoice.server.model;

import java.io.Serializable;

/**
 * Representa um modelo de veículo. Os modelos pertencem sempre a uma marca.
 * 
 * @since 1.0
 */
public final class Modelo implements Serializable {

    private String id;
    private Marca marca;
    private String nome;
    private boolean activo;

    public Modelo() {
        activo = true;
    }

    public Modelo(String id, Marca marca, String nome, boolean activo) {
        this.id = id;
        this.marca = marca;
        this.nome = nome;
        this.activo = activo;
    }

    public Marca getMarca() {
        return marca;
    }

    public void setMarca(Marca marca) {
        this.marca = marca;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public boolean isActivo() {
        return activo;
    }

    public void setActivo(boolean activo) {
        this.activo = activo;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final Modelo other = (Modelo) obj;

        if (this.id != null && this.id.equals(other.id)) {
            return true;
        }

        if (this.marca != other.marca && (this.marca == null || !this.marca.equals(other.marca))) {
            return false;
        }
        if ((this.nome == null) ? (other.nome != null) : !this.nome.equals(other.nome)) {
            return false;
        }
        return true;
    }

    @Override
    public int hashCode() {
        int hash = 3;
        hash = 71 * hash + (this.id != null ? this.id.hashCode() : 0);
        hash = 71 * hash + (this.marca != null ? this.marca.hashCode() : 0);
        hash = 71 * hash + (this.nome != null ? this.nome.hashCode() : 0);
        return hash;
    }
}

File: jai/models/Modelo.php
<?php

/* Modelo.php
 * 
 * This file is part of jAutoInvoice, a car workshop management software.
 * 
 * Copyright (c) 2011, Sérgio Lopes.
 * http://sourceforge.net/projects/jautoinvoice
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * @property int $idModelo
 * @property string $nome
 * @property int $activo
 * @property int $idMarca
 *
 * Relações
 * @property Marca $marca
 * @property Veiculo[] $veiculos
 */
class Modelo extends CActiveRecord {

    /**
     * @return Modelo
     */
    public static function model($className=__CLASS__) {
        return parent::model($className);
    }

    public function tableName() {
        return 'Modelo';
    }

    public function rules() {
        return array(
            array('nome, idMarca', 'required'),
            array('nome', 'length', 'max' => 100),
            // search
            array('idMarca, nome', 'safe', 'on' => 'search'),
        );
    }

    public function relations() {
        return array(
            'marca' => array(self::BELONGS_TO, 'Marca', 'idMarca'),
            'veiculos' => array(self::HAS_MANY, 'Veiculo', 'idModelo'),
        );
    }

    public function attributeLabels() {
        return array(
            'idModelo' => 'ID',
            'nome' => 'Nome',
            'idMarca' => 'Marca'
        );
    }

    /**
     * @return CActiveDataProvider
     */
    public function search() {
        $criteria = new CDbCriteria();
        
        $criteria->order = 'nome';

        $criteria->compare('nome', $this->nome, true);
        $criteria->compare('idMarca', $this->idMarca);
        $criteria->compare('activo', 1);

        return new CActiveDataProvider('Modelo', array('criteria' => $criteria));
    }

}